### ISSUE TITLE
LETKF issue with new MPAS test case

### ISSUE THREAD

**Issue by SamuelDegelia-NOAA** [NOAA-EMC/RDASApp#111]
I am experiencing an issue with running LETKF with the new MPAS test case begin discussed in #91 

LETKF can at least run now when using the 6-h forecast ensemble provided by @guoqing-noaa given that those files have pressure variables. 

However, for some reason the hofx for each ensemble member is the same value for a test I am running with a single surface temperature observation. This leads to no spread and zero increments. 

There is definitely sufficient variability when plotting T at the lowest model level for each member, so the issue is possibly somewhere in MPAS-JEDI instead of in the ensembles. I am currently investigating this problem further.

Quick edit: the problem also occurs when using the sonde observation included with this test case.

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  Could you point to your yaml ?

**Comment by SamuelDegelia-NOAA** [comment]
@TingLei-NOAA the yaml I am currently using to diagnose this problem is `/scratch2/NCEPDEV/stmp3/Samuel.Degelia/mpas_newcase/rrfs-data_mpasjedi_2024052700_share/mpasjedi_letkf_enkfvalidation_oneob.yaml`. That directory is also the run directory for these test experiments.

I have been doing a number of tests to try and figure this out, including using our existing MPAS-JEDI LETKF yaml but with the new ensemble data, and using the yaml that @guoqing-noaa provided for the new test case but with replacing the EnVar sections with LETKF. Both of those tests still have the h(x) problem. 

I also tried using all of the same fix files from the new case but replacing the model data (and associated template, static files) with our old test data. That causes the h(x) calculation to work, so now I am thinking that this is an issue with the new model data. My next step is to add some print statements to the UFO code to see if I can find where the GeoVals for this calculation are coming from.

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  In the jedi.log in your shared dir, there are : 
```
 0: mesonet airTemperature nlocs = 1, nobs = 1, min = 10, max = 10, avg = 10
  0: QC mesonet airTemperature: 1 missing values.
  0: QC mesonet airTemperature: 0 passed out of 1 observations.
```
Did this mean the only one obs was labelled as missing value (and no obs to be used in following analysis)?

**Comment by SamuelDegelia-NOAA** [comment]
@TingLei-NOAA Sorry, I should have included a test with the sonde observation. There is an issue with my single temperature observation that is unrelated to the h(x) issue. See the run directory now that uses `mpasjedi_letkf_enkfvalidation_sonde.yaml` and the `data/obs/rap.t00z.singleobs.ADPUPA.tm00.nc4` observation file. This experiment has the one observation pass QC and  outputs h(x) to `adpupa_hofx.nc4` which is still having the same issue where `hofx0_1`, `hofx0_2`, etc. are all the same value.

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  Looked at hofx for different members with their print-out format  , they showed differences for at least two members (though very small).  Is that possible it just happened the obs was put in a little-spread location? 
I guess, if we use different ensemble member as the background to run a pure hofx.x, we could get the same results. 
For being now, I speculate that the current run with the sounding  is no problem except for the obs is located at a "quiet " region.

**Comment by SamuelDegelia-NOAA** [comment]
@TingLei-NOAA How did you see differences in the hofx values? Mine are showing as the exact same (down to three decimals points): 

```
(eva) [Samuel.Degelia@hfe03 rrfs-data_mpasjedi_2024052700_share]$ python print_hofx.py
Reading adpupa_hofx.nc4...
hofx0_1 = 262.247
hofx0_2 = 262.247
hofx0_3 = 262.247
hofx0_4 = 262.247
hofx0_5 = 262.247
```

I also computed the stddev of temperature at level 33 (closest to 500 mb where the obs is located). While it is small (~0.2 K), this would still suggest that the h(x) should be more different than those output in `adpupa_hofx.nc4`. 

Ensemble stddev of temperature at model level 33 (K): 
<img src="https://github.com/user-attachments/assets/66341135-9986-48d6-affd-7340a00e2fe1" width="60%" height="60%">

**Comment by SamuelDegelia-NOAA** [comment]
Also, here is model level 33 temp plotted for member 1 and member 5. Member 5 is ~3 K warmer at this level. I still think there is an issue with the ensemble hofx for this test case. 

Member 1: 
<img src="https://github.com/user-attachments/assets/96c1995f-c74f-4d05-9098-e346b01b5473" width="60%" height="60%">

Member 2:
<img src="https://github.com/user-attachments/assets/63135bb1-7513-4ac0-b34d-e02317b7d3c4" width="60%" height="60%">

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  Using ncdump, it should be shown as :
```
group: hofx1_1 {
  variables:
        float airTemperature(Location) ;
                airTemperature:_FillValue = -3.368795e+38f ;
  data:

   airTemperature = 262.2462 ;
  } // group hofx1_1
group: hofx1_2 {
  variables:
        float airTemperature(Location) ;
                airTemperature:_FillValue = -3.368795e+38f ;
  data:

   airTemperature = 262.2467 ;
  } // group hofx1_2
  ...
  group: hofx_y_mean_xb1 {
  variables:
        float airTemperature(Location) ;
                airTemperature:_FillValue = -3.368795e+38f ;
  data:

   airTemperature = 262.247 ;
  } // group hofx_y_mean_xb1

```
So, I am not sure if it is an issue with the app or what kind of an issue it would be. 
If a separate hofx run for those members might be helpful .

**Comment by SamuelDegelia-NOAA** [comment]
@TingLei-NOAA I was under the impression that the `hofx1_n` values are the analysis hofx for each member (which are slightly different as you show), while `hofx0_n` are the background hofx (which are all the same). Those analysis differences could be coming from the posterior inflation. But I agree that running a separate hofx run for these members would be helpful - I will work on that. 

Edit: corrected 'o-a' and 'o-b' to 'analysis' and 'background' when referring to hofx.

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  You are right!  I made a mistake regarding hofx1 as the omb. 
Yes. as your results, all omb are actually the same 262.247.  That makes your speculation more convincing.

**Comment by SamuelDegelia-NOAA** [comment]
Thank you for confirming @TingLei-NOAA! Glad that we are on the same page now. I will keep investigating the problem.

**Comment by SamuelDegelia-NOAA** [comment]
It seems that this problem is related to the `filename_template` files listed under either the `static` and `input` streams in `streams.atmosphere`. Changing those files changes the hofx values that are the same for each member. So it seems that hofx is actually being calculated from these static files instead of the ensemble member files defined in the main yaml.

**Comment by guoqing-noaa** [comment]
This may not be related,  do you have any idea what the ensemble stream means?
```
<stream name="ensemble"
        type="input;output"
        precision="single"
        io_type="pnetcdf,cdf5"
        filename_template="ensemble.nc"
        input_interval="none"
        output_interval="none"
        clobber_mode="overwrite">
        <file name="stream_list/stream_list.atmosphere.ensemble"/>
</stream>
```

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  , Good findings.  Looking forward to a clear documentation on how those setups interact with each others.    Is this relevant to that previous " small spread"  issues in the hybrid analysis for this case?

**Comment by SamuelDegelia-NOAA** [comment]
@guoqing-noaa It does not look like we use the ensemble stream for anything right now. Currently in all of our MPAS-JEDI yamls, we use `stream name: background` for **both** the ensemble members and control member. This means that the same set of field (variables) are read in for both sets. But, technically we could use different (fewer) fields for the ensemble than we do for the control member. So the `stream name="ensemble"` gives us that option to read in a different set of variables for the ensemble.

**Comment by SamuelDegelia-NOAA** [comment]
Also, it looks like the problem in this issue was just user error on my end. I did not realize that the ensemble forecasts in `/scratch1/BMC/wrfruc/gge/RDASApp/expr/jedi-assim_test/rrfs-data_mpasjedi_2024052700/Data/ens.czhou.00z` have a different number of vertical levels than the ensemble included in #91. So when I replaced the ensemble members for some tests, I also needed to replace the files from the `static` and `input` streams. I am not sure exactly how that led to this problem and why it didn't throw an error message, but replacing the template files used for these streams now resolves the hofx issue and each member gives a different value.

**Comment by TingLei-NOAA** [comment]
@SamuelDegelia-NOAA  Did you mean, previously, the ensemble members have different vertical levels compared with the background (and corresponding static files), that caused the identical hofx issue?

**Comment by SamuelDegelia-NOAA** [comment]
@TingLei-NOAA I was using a separate set of ensemble forecasts for testing LETKF here in MPAS-JEDI, different from those in #91. The ensemble in #91 are currently cold start files that cannot be used for LETKF because they do not have pressure data. So I used ensemble forecasts from @chunhuazhou that have pressure. These are not included in #91 because of low spread, but I do not think the spread is literally zero like I found here. 

My problem with testing these new ensemble forecasts for use in LETKF was that I did not replace the background and static stream files because I did not realize that these ensemble forecasts had a different number of vertical levels.

**Comment by guoqing-noaa** [comment]
> Also, it looks like the problem in this issue was just user error on my end. I did not realize that the ensemble forecasts in `/scratch1/BMC/wrfruc/gge/RDASApp/expr/jedi-assim_test/rrfs-data_mpasjedi_2024052700/Data/ens.czhou.00z` have a different number of vertical levels than the ensemble included in #91. So when I replaced the ensemble members for some tests, I also needed to replace the files from the `static` and `input` streams. I am not sure exactly how that led to this problem and why it didn't throw an error message, but replacing the template files used for these streams now resolves the hofx issue and each member gives a different value.

Sorry about this. I forgot that Chunhua's case used the GSL version of the MPAS model and has different vertical levels. 
We will update the ensembles as soon as possible. @chunhuazhou

**Comment by SamuelDegelia-NOAA** [comment]
@guoqing-noaa No biggie - I should have checked this. I also would hope that MPAS-JEDI would have some kind of error checking for when the ensemble/background files are on different grids compared to the template stream files.
